lib/src/http_client_request.dart

```diff
@@ -0,0 +1,276 @@
+  /// [ReadDataCallback], [FailedCallabck], [CanceledCallabck] and [SuccessCallabck].
+  /// Callbacks will be called as per sequence of the events.
+  Future<bool> registerCallbacks(ReadDataCallback onReadData,
+      {RedirectReceivedCallback? onRedirectReceived,
+      ResponseStartedCallback? onResponseStarted,
+      FailedCallabck? onFailed,
+      CanceledCallabck? onCanceled,
+      SuccessCallabck? onSuccess}) {
+    if (_cbh._isStreamClaimed) {
+      return Future.error(ResponseListenerException());
+    }
+    final rc = _cbh.registerCallbacks(onReadData, onRedirectReceived,
+        onResponseStarted, onFailed, onCanceled, onSuccess);
+    _startRequest();
+    return rc;
+  }
+
+  /// Returns [Future] of [HttpClientResponse] which can be listened for server response.
+  ///
+  /// Throws [Exception] if callback based api is in use.
+  /// Throws [UrlRequestException] if request can't be initiated.
+  /// Consumable similar to [HttpClientResponse].
+  @override
+  Future<HttpClientResponse> close() {
+    return Future(() {
+      // If callback based API is being used, throw Exception.
+      if (_cbh._callBackCompleter != null) {
+        throw ResponseListenerException();
+      }
+      _startRequest();
+      return HttpClientResponse._(_cbh.stream);
+    });
+  }
+
+  /// Aborts the client connection.
+  ///
+  /// If the connection has not yet completed, the request is aborted
+  /// and closes the [Stream] with onDone callback you may have
+  /// registered. The [Exception] passed to it is thrown and
+  /// [StackTrace] is printed. If there is no [StackTrace] provided,
+  /// [StackTrace.empty] will be shown. If no [Exception] is provided,
+  /// no exception is thrown.
+  /// If the [Stream] is closed, aborting has no effect.
+  void abort([Object? exception, StackTrace? stackTrace]) {
+    // TODO: Migrate abort code
+    throw UnimplementedError();
+  }
+
+  /// Done is same as [close]. A [HttpClientResponse] future that will complete once the response is available.
+  ///
+  /// If an error occurs before the response is available, this future will complete with an error.
+  @override
+  Future<HttpClientResponse> get done => close();
+
+  @override
+  void add(List<int> data) {
+    // TODO: Implement this with POST request
+    throw UnimplementedError();
+  }
+
+  @override
+  void addError(Object error, [StackTrace? stackTrace]) {
+    // TODO: Implement this with POST request
+    throw UnimplementedError();
+  }
+
+  @override
+  Future addStream(Stream<List<int>> stream) {
+    // TODO: Implement this with POST request
+    throw UnimplementedError();
+  }
+
+  @override
+  Future flush() {
+    // TODO: Implement this with POST request
+    throw UnimplementedError();
+  }
+
+  // Implementation taken from `dart:io`.
+  @override
+  void write(Object? object) {
+    final string = '$object';
+    if (string.isEmpty) return;
+    add(encoding.encode(string));
+  }
+
+  // Implementation taken from `dart:io`.
```

Comment:
Just add the doc comment from dart:io as well, that would be more useful for our users using this API.

(Same for other methods.)

Reply:
In dart:io, doc comments are copied from `IOSink`. As, we're also implementing a `IOSink`, the doc comments are getting copied by default. Hence, I'm keeping it blank, just like `dart:io HttpClientRequest`.